### PR TITLE
C4-990 refactor ingestion listener (C4-1005, C4-1006)

### PR DESCRIPTION
**Pull request by dmichaels-harvard**
Initial refactor of main ingestion message loop in preparation for bringing ingestion to Fourfront.  Introduced `@ingestion_message_handler` decorator to help (which is really [the main event](https://github.com/dbmi-bgm/cgap-portal/pull/692/files#diff-b005d244ecd2286f62f1b6457e917f22be1c44bbbd36641139c64208866c9215) here), and created two of these, one for VCF case and one for the non-VCF case (in separate modules). Then the main ingestion loop in ingestion_listener.IngestionListener.run is now [very small]( https://github.com/dbmi-bgm/cgap-portal/pull/692/files#diff-08cd22f5f81dc39b6b5f96379b8e6be13aa97dcc374ebc6af5dbf4519807857cR494).

Question: Should we move all the `ingestion*py` modules to the `ingestion` sub-directory (and the handler ones to a `handlers` sub-directory there)?

### CODE REVIEW

**Comment by netsettler** [comment]
In the main comment, you write:

> Question: Should we move all the ingestion*py modules to the ingestion sub-directory (and the handler ones to a handlers sub-directory there)?

I think so.

File: src/encoded/ingestion_message_handler_decorator.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    handler = _ingestion_message_handlers.get(message.type, _ingestion_message_handlers.get(None))
+    if handler:
+        return handler(message, listener)
+    else:
+        # If NO message handler is registered for the given message type AND if there
+        # is NO default message handler registered then we regard this as a (runtime) error.
+        raise RuntimeError(f"No ingestion message handler defined for ingestion message type: {message.type}")
+
+
+def for_testing_clear_ingestion_message_handlers():
```
Comment (netsettler) [inline src/encoded/ingestion_message_handler_decorator.py:186]: Can we put the verb first? `clear_ingestion_message_handlers_for_testing`? I've been using `_for_testing` a suffix. Plus I thiink it reads better that way.

File: src/encoded/ingestion_message_handler_vcf.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+
+log = structlog.getLogger(__name__)
+
+
+def includeme(config):
+    pass
+
+
+@ingestion_message_handler(ingestion_type=IngestionMessage.TYPE_VCF)
```
Comment (netsettler) [inline src/encoded/ingestion_message_handler_vcf.py:33]: This function dispatches on type_vcf but then later has a test like `vcf_type == "SV":`. I'm not sure what to make of that.

Comment (dmichaels-harvard) [reply src/encoded/ingestion_message_handler_vcf.py:33]: Looks like that's is a difference data point, from the message metadata (uuid actually), not the message "type" (guess I'd say a sub-type of the VCF type).

File: bin/macpoetry-install [hunk 3]
```diff
@@ -1,3 +1,31 @@
...
+
+if [[ `uname -o` == "Darwin" && `uname -m` == "arm64" ]] ; then
+    #
+    # These worked around Mac M1 issues (after the build) running (at least) make deploy1a.
+    #
+    echo "Some more special installs for Mac M1."
+    pip uninstall -y psutil
+    pip install --no-binary :all: psutil
+    pip uninstall -y pytabix
+    pip install pytabix
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply bin/macpoetry-install:13]: That actually doesn't do the trick. Without this step - uninstall and install of pytabix - I get something like this on make deploy1a ...
```ImportError: dlopen(/Users/dmichaels/.pyenv/versions/3.8.13/envs/cg-3.8.13/lib/python3.8/site-packages/tabix.cpython-38-darwin.so, 0x0002): tried: '/Users/dmichaels/.pyenv/versions/3.8.13/envs/cg-3.8.13/lib/python3.8/site-packages/tabix.cpython-38-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/dmichaels/.pyenv/versions/3.8.13/envs/cg-3.8.13/lib/python3.8/site-packages/tabix.cpython-38-darwin.so' (no such file), '/Users/dmichaels/.pyenv/versions/3.8.13/envs/cg-3.8.13/lib/python3.8/site-packages/tabix.cpython-38-darwin.so' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64'))```

[the remark this answers is not in the document]
Comment (netsettler) [reply bin/macpoetry-install:22]: OK

File: src/encoded/ingestion_message_handler_vcf.py (hunk 2 above)
Comment (netsettler) [reply src/encoded/ingestion_message_handler_vcf.py:33]: OK

File: src/encoded/types/file.py [hunk 4]
```diff
@@ -139,7 +139,13 @@ def external_creds(bucket, key, name=None, profile_name=None):
...
                                 aws_access_key_id=os.environ.get('AWS_ACCESS_KEY_ID'),
                                 aws_secret_access_key=os.environ.get('AWS_SECRET_ACCESS_KEY'))
-        token = conn.get_federation_token(Name=name, Policy=json.dumps(policy))
+        token = None
+        try:
+            token = conn.get_federation_token(Name=name, Policy=json.dumps(policy))
+        except Exception as e:
+            if os.environ.get('AWS_SESSION_TOKEN'):
+                raise RuntimeError("Federated session credentials are not allowed here.")
+            raise e
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply src/encoded/types/file.py:148]: Oooh, I see, I misunderstood, thanks for clarifying, will fix up.

**Review by willronchetti — APPROVED** [review]
Looks good, some small feedback

File: Makefile [hunk 5]
```diff
@@ -47,8 +47,9 @@ configure:  # does any pre-requisite installs
 	@#pip install poetry==1.1.9  # this version is known to work. -kmp 11-Mar-2021
 	# Pin to version 1.1.15 for now to avoid this error:
 	#   Because encoded depends on wheel (>=0.29.0) which doesn't match any versions, version solving failed.
-	pip install poetry==1.1.15
-	pip install setuptools==57.5.0 # this version allows 2to3, any later will break -wrr 20-Sept-2021
+	pip install wheel==0.37.1
+	pip install poetry==1.2.2
+	pip install setuptools==57.5.0  # this version allows 2to3, any later will break -wrr 20-Sept-2021
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply Makefile:52]: Yes, true now, thanks.

File: src/encoded/ingestion/ingestion_message_handler_default.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+from .ingestion_message import IngestionMessage
+from .ingestion_message_handler_decorator import ingestion_message_handler
+from .processors import get_ingestion_processor
+
+
+log = structlog.getLogger(__name__)
+
+
+def includeme(config):
+    pass
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply src/encoded/ingestion/ingestion_message_handler_default.py:17]: It seems we do need this; if not there then error on make deploy1a like "AttributeError: module 'encoded.ingestion.ingestion_message_handler_vcf' has no attribute 'includeme'" (from .../site-packages/pyramid/config/__init__.py); and if also remove ingestion_message_handler_vcf from config.include in src/encoded/__init__.py then error on make deploy2 like "RuntimeError: No ingestion message handler defined for ingestion message type: metadata_bundle" ... It seems this module which needs to be loaded for (now dynamic, dispatch-based) message handling needs to be loaded "manually" using this mechanism.

File: src/encoded/ingestion/ingestion_message_handler_decorator.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        # will ASSUME in this case is a RAW (dict) message from which we create an IngestionMessage.
+        message = IngestionMessage(message)
+    # Get the handler for this message type, or the default handler of none specifically found.
+    handler = _ingestion_message_handlers.get(message.type, _ingestion_message_handlers.get(None))
+    if handler:
+        return handler(message, listener)
+    else:
+        # If NO message handler is registered for the given message type AND if there
+        # is NO default message handler registered then we regard this as a (runtime) error.
+        raise RuntimeError(f"No ingestion message handler defined for ingestion message type: {message.type}")
```
[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply src/encoded/ingestion/ingestion_message_handler_decorator.py:183]: Not sure what you mean by the entire message ...

[the remark this answers is not in the document]
Comment (willronchetti) [reply src/encoded/ingestion/ingestion_message_handler_decorator.py:183]: I just mean the `message` object you pulled from SQS

[the remark this answers is not in the document]
Comment (dmichaels-harvard) [reply src/encoded/ingestion/ingestion_message_handler_decorator.py:183]: Oh I see, the `message` (not the "message"). Sounds good.

**Pull request merged**